src/psyclone/domain/lfric/lfric_types.py

```diff
@@ -48,206 +49,298 @@ class LFRicTypes:
     '''This class implements a singleton that manages LFRic types.
     Using the 'call' interface, you can query the data type for
     LFRic types, e.g.:
-    lfric_types = LFRicTypes()
-    rvfdds = lfric_types("RealVectorFieldDataDataSymbol")
+
+    >>> from psyclone.domain.lfric import LFRicTypes
+    >>> num_dofs_class = LFRicTypes("NumberOfUniqueDofsDataSymbol")
+    >>> my_var = num_dofs_class("my_num_dofs")
+    >>> print(my_var.name)
+    my_num_dofs
+
+    It uses the __new__ function to implement the access to the internal
+    dictionary. This is done to minimise the required code for getting a
+    value, e. g. compared with ``LFRicTypes.get()("something")``, or
+    ``LFRicType.get("something")``.
 
     '''
 
     # Class variable to store the singleton instance
     _instance = None
 
+    # Class variable to store the mapping of names to the various objects
+    # (classes and various instances) managed by this class.
+    _name_to_class = {}
+
     # ------------------------------------------------------------------------
-    def __new__(cls):
-        '''Implement a singleton - only one instance will ever be created.
+    def __new__(cls, name):
+        '''The class creation function __new__ is used to actually provide
+        the object the user is querying. It is well documented that __new__
+        can return a different instance. This function will first make sure
+        that the static internal dictionary is initialised (so it acts like
+        a singleton in that regard). Then it will return the value the user
+        asked for.
 
-        :returns: the singleton instance of LFRicTypes.
-        :rtype: :py:class:`psyclone.domain.lfric.LFRicTypes`
+        :param str name: the name to query for.
+
+        :returns: the corresponding object, which can be a class or an \
+            instance.
+        :rtype: object (various types)
 
         '''
-        if LFRicTypes._instance is None:
-            # Return a new instance. The constructor will set _instance
-            # in this case
-            return super().__new__(cls)
+        if not LFRicTypes._name_to_class:
+            LFRicTypes.init()
 
-        # Return the existing instance, in which case the constructor will
-        # not re-initialise the internal data structures
-        return LFRicTypes._instance
+        value = LFRicTypes._name_to_class.get(name, None)
```

Comment:
Perhaps a `try...except KeyError` would be nicer here?

Reply:
Done. I prefer `get(..., None)` since I think it causes less overhead: a `try` means that python has to setup some exception handlers, while with a `get` this is not ncessary and I assume it would therefore be faster. But I admit it won't make any noticeable speed difference :)
OK, I got curious if this is actually true:
https://stackoverflow.com/questions/2522005/cost-of-exception-handlers-in-python
So, if the exception is not triggered, an `if` statement is a tiny bit faster. If the exception is triggered, an `if` test is significantly faster (>10x). But again, it doesn't matter here.

Reply:
Oh, and it only saved one line in the end. Oh well.